action/protocol/staking/contractstaking/indexer.go

```diff
@@ -82,3 +105,110 @@ func (s *Indexer) BucketTypes() ([]*BucketType, error) {
 	}
 	return bts, nil
 }
+
+// PutBlock puts a block into indexer
+func (s *Indexer) PutBlock(ctx context.Context, blk *block.Block) error {
+	// new dirty cache for this block
+	// dirty is the cache to update during event processing, which will be merged to clean cache after all events are processed.
+	// if errors occur during event processing, dirty cache will be discarded.
+	dirty := newContractStakingDirty(s.cache)
+	dirty.PutHeight(blk.Height())
```

Comment:
convert these to `cache's` func

Reply:
```
cache.NewWorkingset(blk.Height())
```

Reply:
instead of indexer --> cache, dirty
indexer --> cache
let cache handle --> dirty inside

Reply:
make the indexer unaware of the `dirty`